group/assignment6/RearrangeCars.cpp

```diff
@@ -3,155 +3,218 @@
-				sequenceOfMoves.push_back(Move(currentStates[emptySlotId], make_pair(currCarSlotId, emptySlotId)));
-				bruteForceAllSequence(currentStates, endStates, carPosition, currCarSlotId, sequenceOfMoves, sequencesOfMoves, usedPositions, constraints);
-				sequenceOfMoves.pop_back();
-				usedPositions.erase(usedPositions.find(carPosition));
-			}
-
-			// we want to swap again and return to our currentStates for next try
-			moveCar(currentStates, carPosition, emptySlotId, currCarSlotId);
-		}
-	}
+void
+RearrangeCars::bruteForceAllSequence(vector<int> &currentStates, const vector<int> &endStates, vector<int> &carPosition,
+                                     int emptySlotId, vector<Move> &sequenceOfMoves,
+                                     vector<vector<Move>> &sequencesOfMoves, set<vector<int>> &usedPositions,
+                                     const vector<set<int>> *constraints) {
+      int numOfCarsOnDisiredPosition = numberOfCarsOnTheirDesiredPositions(carPosition, endStates);
+      if (numOfCarsOnDisiredPosition == carPosition.size()) {
+            sequencesOfMoves.push_back(sequenceOfMoves);
+            return;
+      }
+      for (int currCarSlotId = 0; currCarSlotId < currentStates.size(); currCarSlotId++) {
+            if (currCarSlotId != emptySlotId && validMove(currentStates[currCarSlotId], emptySlotId, constraints)) {
+
+                  // we swap car on our currCarSlotId and now this car on emptySlotId position
+                  moveCar(currentStates, carPosition, currCarSlotId, emptySlotId);
+
+                  if (usedPositions.find(carPosition) == usedPositions.end()) {
+                        usedPositions.insert(carPosition);
+                        sequenceOfMoves.push_back(Move(currentStates[emptySlotId], make_pair(currCarSlotId, emptySlotId)));
+                        bruteForceAllSequence(currentStates, endStates, carPosition, currCarSlotId,
+                                              sequenceOfMoves, sequencesOfMoves, usedPositions, constraints);
+                        sequenceOfMoves.pop_back();
+                        usedPositions.erase(usedPositions.find(carPosition));
+                  }
+
+                  // we want to swap again and return to our currentStates for next try
+                  moveCar(currentStates, carPosition, emptySlotId, currCarSlotId);
+            }
+      }
 }
 
-void RearrangeCars::generateAllSequencesOfMoves(const vector<int> &endStates, vector<vector<Move>> &sequencesOfMoves, const vector<set<int>> *constraints) {
-	vector<int> currentStates = this->startStates;
-	vector<int> carPosition = findCarPositions(currentStates);
+bool RearrangeCars::generateAllSequencesOfMoves(const vector<int> &endStates, vector<vector<Move>> &sequencesOfMoves,
+                                                const vector<set<int>> *constraints) {
+      sequencesOfMoves.clear();
+      if (!isInputValid(this->startStates) || !isInputValid(endStates)) {

```

Comment:
I believe it is a bit overcomplicated, don't you think so? I don't really understand why should we call isInputValid several times and why don't we put constraints as the argument of input validation. Could you explain this for me, please? I thought it would be something like
`if (!isInputValid(endStates, constraints)) 
    return false;`

Reply:
You're absolutely right, my bad
Fixing